File: client/ios/DivKit/Actions/CopyToClipboardActionHandler.swift
import Foundation

#if os(iOS)
import UIKit
#endif

final class CopyToClipboardActionHandler {
  func handle(
    _ action: DivActionCopyToClipboard,
    context: DivActionHandlingContext
  ) {
    #if os(iOS)
    let pasteboard = UIPasteboard.general
    switch action.content {
    case let .contentText(text):
      if let text = text.resolveValue(context.expressionResolver) {
        pasteboard.string = text
      }
    case let .contentUrl(url):
      if let url = url.resolveValue(context.expressionResolver) {
        pasteboard.url = url
      }
    }
    #endif
  }
}

File: client/android/compose/src/main/kotlin/com/yandex/div/compose/actions/CopyToClipboardActionHandler.kt
package com.yandex.div.compose.actions

import android.content.ClipboardManager
import android.content.Context
import androidx.core.content.getSystemService
import com.yandex.div.compose.DivReporter
import com.yandex.div.internal.actions.resolveClipData
import com.yandex.div2.DivActionCopyToClipboard
import javax.inject.Inject

internal class CopyToClipboardActionHandler @Inject constructor(
    private val context: Context,
    private val reporter: DivReporter,
) {

    fun handle(context: DivActionHandlingContext, action: DivActionCopyToClipboard) {
        val clipboard = this.context.getSystemService<ClipboardManager>()
        if (clipboard == null) {
            reporter.reportError("Failed to access clipboard manager")
            return
        }

        clipboard.setPrimaryClip(action.content.resolveClipData(context.expressionResolver))
    }
}
